src/awkward/contents/unionarray.py

```diff
@@ -1574,13 +1574,23 @@ def _remove_structure(
         self, backend: Backend, options: RemoveStructureOptions
     ) -> list[Content]:
         out = []
-        for i in range(len(self._contents)):
-            index = self._index[self._tags.data == i]
-            out.extend(
-                self._contents[i]
-                ._carry(index, False)
+
+        # typetracer
+        if not self._backend.nplike.known_data:
+            self._touch_data(recursive=False)
+            # just flatten, ignore order, tags, index
+            for c in self._contents:
+                out.extend(c._remove_structure(backend, options))
+            return out
+
+        # backends with concrete data
+        for i in range(self._tags.length):
+            content = (
+                self._contents[self._tags.data[i]]
+                ._carry(ak.index.Index(self._index.data[i]), False)
                 ._remove_structure(backend, options)
             )
+            out.extend(content)
```

Comment:
@pfackeldey this is a correct solution as far as structure goes, although allocates repeatedly according to the tags which is not ideal. 

Would you be able to leave a `# FIXME: ` note here (probably just commit to `main`) that we can optimise this by building a `UnionArray` over the original naive loop to re-order the result? That might not be trivial to do, but we should definitely remark upon it!